libazureinit/src/health.rs

```diff
@@ -0,0 +1,578 @@
+use crate::error::Error;
+use crate::http;
+
+#[derive(Debug)]
+enum ProvisioningState {
+    Ready,
+    NotReady,
+    InProgress,
+}
+
+#[derive(Debug)]
+enum ProvisioningSubStatus {
+    ProvisioningFailed,
+    ProvisioningInProgress,
+}
+
+impl std::fmt::Display for ProvisioningState {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = match self {
+            ProvisioningState::Ready => "Ready",
+            ProvisioningState::NotReady => "NotReady",
+            ProvisioningState::InProgress => "InProgress",
+        };
+        write!(f, "{s}")
+    }
+}
+
+impl std::fmt::Display for ProvisioningSubStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let s = match self {
+            ProvisioningSubStatus::ProvisioningFailed => "ProvisioningFailed",
+            ProvisioningSubStatus::ProvisioningInProgress => {
+                "ProvisioningInProgress"
+            }
+        };
+        write!(f, "{s}")
+    }
+}
+
+#[derive(Serialize)]
+pub struct ReportableError {
+    result: &'static str,
+    reason: String,
+    documentation_url: &'static str,
+    #[serde(flatten)]
+    supporting_data: HashMap<String, String>,
+}
+
+impl std::fmt::Display for ReportableError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.reason)
+    }
+}
+
+impl ReportableError {
+    pub fn new(reason: impl Into<String>) -> Self {
+        ReportableError {
+            result: "error",
+            reason: reason.into(),
+            documentation_url: "https://aka.ms/linuxprovisioningerror",
+            supporting_data: HashMap::new(),
+        }
+    }
+
+    /// Adds a key-value pair to the supporting data for the report.
+    pub fn with_supporting_data(mut self, key: &str, value: &str) -> Self {
+        self.supporting_data
+            .insert(key.to_string(), value.to_string());
+        self
+    }
+
+    /// Serializes the reportable error as a pipe-delimited KVP entry.
+    pub fn as_encoded_report(
+        &self,
+        vm_id: &str,
+        agent: &str,
+        timestamp: &DateTime<Utc>,
+        _pps_type: &str,
+    ) -> String {
+        let mut data = vec![
+            "result=error".to_string(),
+            format!("reason={}", self.reason),
+            format!("agent={}", agent),
+        ];
+        for (k, v) in &self.supporting_data {
+            data.push(format!("{k}={v}"));
+        }
+        data.push("pps_type=None".to_string());
+        data.push(format!("vm_id={vm_id}"));
+        data.push(format!("timestamp={}", timestamp.to_rfc3339()));
+        data.push(format!("documentation_url={}", self.documentation_url));
+        encode_report(&data)
+    }
+
+    /// Constructor for “failure to find DHCP interface”.
+    pub fn dhcp_interface_not_found(duration_secs: f64) -> Self {
+        let mut err = ReportableError::new("failure to find DHCP interface");
```

Comment:
we can lose these constructors and just have teh caller do them unless perhaps reused in multiple places